kusto/kcsb.go

```diff
@@ -0,0 +1,354 @@
+	requireNonEmpty(authorityId, authorityID)
+	kcsb.applicationClientId = appId
+	kcsb.authorityId = authorityID
+
+	kcsb.applicationCertificate = certificate
+	kcsb.applicationCertificateThumbprint = thumprint
+	kcsb.sendCertificateChain = sendCertChain
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will authenticate with AAD application and an application token.
+func (kcsb connectionStringBuilder) WithApplicationToken(appId string, appToken string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationToken, appToken)
+	kcsb.applicationToken = appToken
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will use existing authenticated az cli profile password.
+func (kcsb connectionStringBuilder) WithAzCli() connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.azcli = true
+	return kcsb
+}
+
+/*
+Creates a KustoConnection string builder that will authenticate with AAD application, using
+  an application token obtained from a Microsoft Service Identity endpoint. An optional user
+  assigned application ID can be added to the token.
+*/
+func (kcsb connectionStringBuilder) WithManagedServiceID(clientID string, resId string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.msiAuthentication = true
+	if !isEmpty(clientID) {
+		kcsb.managedServiceIdentity = clientID
+	} else if !isEmpty(resId) {
+		kcsb.managedServiceIdentity = resId
+	}
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) WithInteractiveLogin(clientID string, authorityID string, redirectURL string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	if !isEmpty(clientID) {
+		kcsb.applicationClientId = clientID
+	}
+	if !isEmpty(authorityID) {
+		kcsb.authorityId = authorityID
+	}
+	if !isEmpty(redirectURL) {
+		kcsb.redirectURL = redirectURL
+	}
+	kcsb.interactiveLogin = true
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) AttachClientOptions(options *azcore.ClientOptions) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	if options == nil {
+		kcsb.clientOptions = options
+	}
+	return kcsb
+}
+
+// Method to be used for generating TokenCredential
+func (kcsb connectionStringBuilder) getTokenProvider() (*tokenProvider, error) {
+	tkp := &tokenProvider{}
+	tkp.dataSource = kcsb.dataSource
+	if kcsb.interactiveLogin {
+		inOps := &azidentity.InteractiveBrowserCredentialOptions{}
+
+		if !isEmpty(kcsb.applicationClientId) {
+			inOps.ClientID = kcsb.applicationClientId
+		}
+		if !isEmpty(kcsb.authorityId) {
+			inOps.TenantID = kcsb.authorityId
+		}
+		if kcsb.clientOptions != nil {
+			inOps.ClientOptions = *kcsb.clientOptions
+		}
+		cred, err := azidentity.NewInteractiveBrowserCredential(inOps)
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Interactive Login. Error: %s", err)
+		}
+		tkp.tokenCred = cred
```

Comment:
Return and then you can skip the else

Reply:
TThis function sometimes early returns and sometimes doesn't. Please make it consistantly early return and then you can get rid of elses and indentations